viamrobotics/rdk
# RSDK-4810 integrate StopPlan and MoveOnGlobeNew with ExecutionId system

First Attempt: https://github.com/viamrobotics/rdk/pull/3010
Second Attempt: https://github.com/viamrobotics/rdk/pull/3238

### Tickets:
https://viam.atlassian.net/browse/RSDK-4810
https://viam.atlassian.net/browse/RSDK-4811
https://viam.atlassian.net/browse/RSDK-4812
https://viam.atlassian.net/browse/RSDK-4814

- change moveReqeusts of type MoveOnGlobe to include steps which are smuggled geoposes.
- create services/motion/builtin/state package to manage the state of the builtin motion service's executions, which supports starting executions, stopping executions by resource, stopping all executions from a motion/builtin/state struct, recording succeeded & failed replans, recording succeeded & failed ends to executions, recording stopped executions, and recording plan status updates, all such that data can be updated & retrieved efficiently.
- change motion/builtin.MoveOnGlobeNew to call state.StartExecution
- change motion/builtin.StopPlan to call state.StopExecutionByResource
- change motion/builtin.ListPlanStatuses to call state.ListPlanStatuses
- change motion/builtin.PlanHistory to call state.PlanHistory
- change services/motion/builtin to create a new motion/builtin/state on reconfigure & stop the previous state if one exists (to stop any executions from prior to reconfiguration)
- change ptgkinematic base to stop the base with a 5 second timeout if the context is cancelled


Manual Testing:
Tested by running the script below using a base & movement sensor simulated in gazebo.

1. Result of running MoveOnGlobeNew & then querying GetPlan to see the plans generated during the execution. In the video below motion replanned twice, resulting in 3 plans:

https://github.com/viamrobotics/rdk/assets/5927876/9d8e41ba-8f3c-4214-8235-9e950f6776f3

2. Result of calling StopPlan:
https://github.com/viamrobotics/rdk/assets/5927876/8931d70b-b43a-4f8e-9fa8-27a6dee6fa49



```go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"os/signal"
	"strconv"
	"sync"
	"syscall"
	"time"

	"github.com/edaniels/golog"
	"github.com/google/uuid"
	geo "github.com/kellydunn/golang-geo"
	v1 "go.viam.com/api/service/motion/v1"
	"go.viam.com/rdk/components/base"
	"go.viam.com/rdk/components/movementsensor"
	"go.viam.com/rdk/robot/client"
	"go.viam.com/rdk/services/motion"
	"go.viam.com/utils/rpc"
)

type state struct {
	shutdown    chan os.Signal
	done        chan struct{}
	cancelFn    context.CancelFunc
	m           motion.Service
	req         motion.MoveOnGlobeReq
	executionID string
	heading     float64
	position    *geo.Point
	dest        *geo.Point
	waitGroup   sync.WaitGroup
}

func main() {
	shutdown := make(chan os.Signal, 1)
	done := make(chan struct{}, 1)
	signal.Notify(shutdown, syscall.SIGINT, syscall.SIGTERM)
	if len(os.Args) != 3 {
		fmt.Fprintln(os.Stderr, "usage: <cmd> lat lng")
		return
	}
	logger := golog.NewLoggerForGCP("client")
	lat, err := strconv.ParseFloat(os.Args[1], 64)
	if err != nil {
		fmt.Fprintf(os.Stderr, "lat not able to be parsed %s\n", err)
	}

	lng, err := strconv.ParseFloat(os.Args[2], 64)
	if err != nil {
		fmt.Fprintf(os.Stderr, "lng not able to be parsed %s\n", err)
	}

	dest := geo.NewPoint(lat, lng)

	robot, err := client.New(
		context.Background(),
		"",
		logger,
		client.WithDialOptions(rpc.WithEntityCredentials(
			"",
			rpc.Credentials{
				Type:    rpc.CredentialsTypeAPIKey,
				Payload: "",
			})),
	)
	if err != nil {
		logger.Fatal(err)
	}

	defer robot.Close(context.Background())

	merged, err := movementsensor.FromRobot(robot, "merged")
	if err != nil {
		logger.Fatal(err)
		return
	}
	m, err := motion.FromRobot(robot, "builtin")
	if err != nil {
		logger.Fatal(err)
		return
	}
	position, _, err := merged.Position(context.Background(), nil)
	if err != nil {
		logger.Fatal(err)
		return
	}

	heading, err := merged.CompassHeading(context.Background(), nil)
	if err != nil {
		logger.Fatal(err)
		return
	}

	req := motion.MoveOnGlobeReq{
		ComponentName:      base.Named("base"),
		Destination:        dest,
		MovementSensorName: movementsensor.Named("merged"),
		Heading:            math.NaN(),
		MotionCfg: &motion.MotionConfiguration{
			PositionPollingFreqHz: 1,
			ObstaclePollingFreqHz: 2,
			PlanDeviationMM:       2600,
			LinearMPerSec:         0.3,
			AngularDegsPerSec:     20,
		},
		Extra: map[string]interface{}{"smooth_iter": 20},
	}
	ctx, cancelFn := context.WithCancel(context.Background())
	executionID, err := m.MoveOnGlobeNew(ctx, req)
	if err != nil {
		cancelFn()
		logger.Fatal(err)
		return
	}
	state := &state{
		shutdown:    shutdown,
		done:        done,
		cancelFn:    cancelFn,
		m:           m,
		req:         req,
		executionID: executionID,
		heading:     heading,
		position:    position,
		dest:        dest,
	}
	poll(ctx, state, logger)
	state.waitGroup.Wait()
}

func poll(ctx context.Context, s *state, logger golog.Logger) {
	s.waitGroup.Add(1)
	go func() {
		defer s.waitGroup.Done()
		select {
		case <-s.shutdown:
		case <-s.done:
		}
		s.cancelFn()
		if err := s.m.StopPlan(context.Background(), motion.StopPlanReq{ComponentName: s.req.ComponentName}); err != nil {
			logger.Errorf("error stopping plan: %s", err)
		}
	}()
	fmt.Fprintf(os.Stderr, "executionID: %v\n", s.executionID)
	defer func() {
		s.done <- struct{}{}
	}()
	for {
		lps, err := s.m.ListPlanStatuses(ctx, motion.ListPlanStatusesReq{})
		if err != nil {
			logger.Error(err.Error())
			return
		}
		var found bool
		for _, ps := range lps {
			if ps.ExecutionID.String() == s.executionID {
				found = true
				if ps.Status.State != motion.PlanStateInProgress {
					ph, err := s.m.PlanHistory(ctx, motion.PlanHistoryReq{
						ComponentName: s.req.ComponentName,
						ExecutionID:   uuid.MustParse(s.executionID),
					})
					if err != nil {
						logger.Error(err.Error())
						return

					}
					pwsProto := make([]*v1.PlanWithStatus, 0, len(ph))
					for _, pws := range ph {
						pwsProto = append(pwsProto, pws.ToProto())
					}
					b, err := json.Marshal(map[string]interface{}{
						"start_heading":  s.heading,
						"start_position": map[string]float64{"lat": s.position.Lat(), "lng": s.position.Lng()},
						"goal_position":  map[string]float64{"lat": s.dest.Lat(), "lng": s.dest.Lng()},
						"plan_history":   pwsProto,
					})
					if err != nil {
						logger.Error(err.Error())
						return
					}
					fmt.Println(string(b))
					return
				}
				fmt.Fprintf(os.Stderr, "state: %s\n", ps.Status.State)
				break
			}
		}

		if !found {
			logger.Error("execution id not found in ListPlanStatuses response")
			return
		}
		time.Sleep(time.Second)
	}
}
```
